PXC-3353 Modify error handling in Garbd when donor crashes during SST or when an invalid donor name is passed to it

venkatesh-prasad-v opened the pull request:
https://jira.percona.com/browse/PXC-3353

Problem
-------
When the donor node is killed while SST is in progress or when an
invalid donor is passed, the garbd hung forever waiting for the receiver
queue to get empty.

Analysis
--------
In both of the above cases, GCS (group_find_node_by_name()) considered
these errors as EHOSTUNREACH (No Route to Host), and triggers the
shutdown of garbd.

While the garbd main thread performs cleanup, the receiver thread would
still be queuing the received actions to the receiver queue until it
exits. So, this would result in the receiver queue to have a few
received actions which shall never be applied by the applier/main
thread.

When the main thread is performing the cleanup of the GCS
(gcs_destroy()), it sees the receiver queue to be not empty
(fifo_flush()) and thus ends up waiting on the conditional variable for
there will be no one to signal as all threads are already joined.

The garbd log will have

```
2020-09-16 09:18:19.664  INFO: recv_thread() joined.
2020-09-16 09:18:19.664  INFO: Closing replication queue.
2020-09-16 09:18:19.664  INFO: Closing slave action queue.
2020-09-16 09:18:19.664  WARN: Waiting for 2 items to be fetched.
... hangs ...
```
Stacktrace of the hang

```
(gdb) bt
futex_wait_cancelable
__pthread_cond_wait_common
__pthread_cond_wait
fifo_flush
gu_fifo_destroy
gcs_destroy
garb::Gcs::~Gcs
garb::RecvLoop::RecvLoop
garb::main
main
```

Fix
---
Modified the error handling to empty the contents of the receiver queue
when garbd is exiting.

Additionally, this commit
1. Modifies the cmake/dev-helper.cmake to build the galera library in debug
    mode when the server is built with -DWITH_GALERA_DEV
2. Updates the galera pointer.

Testing
-------
Jenkins: https://pxc.cd.percona.com/view/PXC%208.0/job/pxc-8.0-param/90/testReport/

Summary: The test `galera.galera_garbd_sst_failure` failed because of the missing `have_debug_sync.inc`. Other test failures are sporadic in nature.

kennt-percona commented on cmake/dev-helper.cmake:
```diff
@@ -7,7 +7,7 @@ if(WITH_GALERA_DEV)
 
   # Add a custom target for later refreshes
-  ADD_CUSTOM_TARGET(galera ALL scons -j ${CPU_COUNT} tests=0
+  ADD_CUSTOM_TARGET(galera ALL scons -j ${CPU_COUNT} debug=1 tests=0
```
does this mean that it always builds debug?

venkatesh-prasad-v replied:
@kennt-percona  Yes. Without this, many galera MTR test cases skipped locally because of missing debug sync in galera.

kennt-percona replied:
I guess my question is whether this should be dependent on the CMAKE_BUILD_TYPE or is debug always enabled (even for release builds).

venkatesh-prasad-v replied:
Good point! IIUC, this option was introduced to simplify the galera development for developers and I believe it is not being used in our Jenkins as well and we are the only ones who are using this option for building galera locally. So, as of now, I feel it is safe to build it in debug mode always.

But, given the advantages we get with building as per the CMAKE_BUILD_TYPE option, then definitely I agree with you. I will change the script accordingly and will update it.

Thank you for suggesting the improvement.

kamil-holubicki commented on mysql-test/suite/galera/t/galera_garbd_invalid_donor_name.test:
```diff
@@ -0,0 +1,94 @@
+# ==== Purpose ====
+#
+# This test verifies that garbd shuts down when when donor node is killed while
```
Is this description OK?

venkatesh-prasad-v replied:
Thank you for pointing it out. Fixed in the next revision.

kamil-holubicki commented on mysql-test/suite/galera/t/galera_garbd_invalid_donor_name.test:
```diff
@@ -0,0 +1,94 @@
+
+--source include/wait_proc_to_finish.inc
+
+# node1 has wsrep_cluster_address=gcomm:// in my.cnf, and since is not the last
```
Why we touch node_1 here? It is still running. We shut down only node_2 and used garbd in place of it. Now we want to start node_2 again.

venkatesh-prasad-v replied:
Right! This came as a side-effect of copy-paste and is not needed. Will fix this in next revision.

kamil-holubicki commented on mysql-test/suite/galera/t/galera_garbd_invalid_donor_name.test:
```diff
@@ -0,0 +1,94 @@
+
+# Add test suppressions
+--connection node_1
+CALL mtr.add_suppression("Member .*garb.* requested state transfer from 'invalid_donor', but it is impossible to select State Transfer donor: No route to host");
```
Suppressions at the beginning of the test?

venkatesh-prasad-v replied:
In fact, the MTR suppressions can put anythere in the test, but it is recommended to put it either near the top, or just after the action that may result in the warning.

So far the current case is concerned, I thought it would be more appropriate if we place it right after the action so that it justifies the need for the suppression.

kamil-holubicki commented on mysql-test/suite/galera/t/galera_garbd_sst_failure.test:
```diff
@@ -0,0 +1,121 @@
+--let $assert_file= $GARBD_LOG_FILE
+--source include/assert_grep.inc
+
+# Add test suppressions
```
Shouldn't suppressions be set at the beginning?

venkatesh-prasad-v replied:
Same as above.

Commit:
PXC-3353 Modify error handling in Garbd when donor crashes during SST or
         when an invalid donor name is passed to it

https://jira.percona.com/browse/PXC-3353

Problem
-------
When the donor node is killed while SST is in progress or when an
invalid donor is passed, the garbd hung forever waiting for the receiver
queue to get empty.

Analysis
--------
In both of the above cases, GCS (group_find_node_by_name()) considered
these errors as EHOSTUNREACH (No Route to Host), and triggers the
shutdown of garbd.

While the garbd main thread performs cleanup, the receiver thread would
still be queuing the received actions to the receiver queue until it
exits. So, this would result in the receiver queue to have a few
received actions which shall never be applied by the applier/main
thread.

When the main thread is performing the cleanup of the GCS
(gcs_destroy()), it sees the receiver queue to be not empty
(fifo_flush()) and thus ends up waiting on the conditional variable for
there will be no one to signal as all threads are already joined.

The garbd log will have

2020-09-16 09:18:19.664  INFO: recv_thread() joined.
2020-09-16 09:18:19.664  INFO: Closing replication queue.
2020-09-16 09:18:19.664  INFO: Closing slave action queue.
2020-09-16 09:18:19.664  WARN: Waiting for 2 items to be fetched.
... hangs ...

Stacktrace of the hang

(gdb) bt
futex_wait_cancelable
__pthread_cond_wait_common
__pthread_cond_wait
fifo_flush
gu_fifo_destroy
gcs_destroy
garb::Gcs::~Gcs
garb::RecvLoop::RecvLoop
garb::main
main

Fix
---
Modified the error handling to empty the contents of the receiver queue
when garbd is exiting.

Additionally, this commit
1. Modifies the cmake/dev-helper.cmake to build the galera library as
   per the CMAKE_BUILD_TYPE mode when the server is built with
   -DWITH_GALERA_DEV
2. Updates the galera pointer.

kamil-holubicki reviewed (approved):
lgtm
